modules/images/image-loading-optimization/optimization.php

```diff
@@ -0,0 +1,174 @@
+	$slug = ilo_get_url_metrics_slug( ilo_get_normalized_query_vars() );
+	$post = ilo_get_url_metrics_post( $slug );
+
+	// No URL metrics are present, so there's nothing we can do.
+	if ( ! $post ) {
+		return $buffer;
+	}
+
+	$url_metrics = ilo_parse_stored_url_metrics( $post );
+
+	$breakpoint_max_widths                   = ilo_get_breakpoint_max_widths();
+	$url_metrics_grouped_by_breakpoint       = ilo_group_url_metrics_by_breakpoint( $url_metrics, $breakpoint_max_widths );
+	$lcp_elements_by_minimum_viewport_widths = ilo_get_lcp_elements_by_minimum_viewport_widths( $url_metrics_grouped_by_breakpoint );
+	$all_breakpoints_have_url_metrics        = count( array_filter( $url_metrics_grouped_by_breakpoint ) ) === count( $breakpoint_max_widths ) + 1;
+
+	// TODO: Handle case when the LCP element is not an image at all, but rather a background-image.
+	// Prepare to set fetchpriority attribute on the image when all breakpoints have the same LCP element.
+	if (
+		// All breakpoints share the same LCP element (or all have none at all).
+		1 === count( $lcp_elements_by_minimum_viewport_widths )
+		&&
+		// The breakpoints don't share a common lack of an LCP element.
+		! in_array( false, $lcp_elements_by_minimum_viewport_widths, true )
+		&&
+		// All breakpoints have URL metrics being reported.
+		$all_breakpoints_have_url_metrics
+	) {
+		$common_lcp_element = current( $lcp_elements_by_minimum_viewport_widths );
+	} else {
+		$common_lcp_element = null;
+	}
+
+	// Walk over all IMG tags in the document and ensure fetchpriority is set/removed, and gather IMG attributes for preloading.
+	$processor = new ILO_HTML_Tag_Processor( $buffer );
+	foreach ( $processor->open_tags() as $tag_name ) {
+		if ( 'IMG' !== $tag_name ) {
+			continue;
+		}
+
+		// Ensure the fetchpriority attribute is set on the element properly.
+		if ( $common_lcp_element && $processor->get_breadcrumbs() === $common_lcp_element['breadcrumbs'] ) {
+			if ( 'high' === $processor->get_attribute( 'fetchpriority' ) ) {
+				$processor->set_attribute( 'data-ilo-fetchpriority-already-added', true );
+			} else {
+				$processor->set_attribute( 'fetchpriority', 'high' );
+				$processor->set_attribute( 'data-ilo-added-fetchpriority', true );
+			}
+		} elseif ( $all_breakpoints_have_url_metrics && $processor->get_attribute( 'fetchpriority' ) ) {
+			// Note: The $all_breakpoints_have_url_metrics condition here allows for server-side heuristics to
+			// continue to apply while waiting for all breakpoints to have metrics collected for them.
+			$processor->set_attribute( 'data-ilo-removed-fetchpriority', $processor->get_attribute( 'fetchpriority' ) );
+			$processor->remove_attribute( 'fetchpriority' );
+		}
+
+		// Capture the attributes from the LCP elements to use in preload links.
+		foreach ( $lcp_elements_by_minimum_viewport_widths as &$lcp_element ) {
+			if ( $lcp_element && $lcp_element['breadcrumbs'] === $processor->get_breadcrumbs() ) {
```

Comment:
Let's optimize this, as a `foreach` for every tag just to find the right entry isn't great and easy to resolve. We can either update `$lcp_elements_by_minimum_viewport_widths` or create a copy from it in another variable where the entries are keyed by the `breadcrumbs` value.

Alternatively, since we want to update `$lcp_elements_by_minimum_viewport_widths` here, we could make the new variable a simple lookup, where keys are the `breadcrumbs` values and values are the respective indexes in `$lcp_elements_by_minimum_viewport_widths`.

That way we can avoid this `foreach` and simply check for `isset( $new_var[ $processor->get_breadcrumbs() ] )`.

Reply:
Done in 74145a460146521da14bfdc2f769a5faccf22f29.

Since breadcrumbs are an array, they can't be used as an array key. So I have to convert to a string first. This string I intend to be reusing later as mentioned in https://github.com/WordPress/performance/pull/884#discussion_r1412402227

Reply:
Thanks, that looks great. Good point on having to transform breadcrumbs to a string.